File: ReverieCli-ui/src/thinking-tool.ts
/**
 * Shared vocabulary for the experimental Thinking Tool.
 *
 * Mirrors `reverie/thinking_tool.py` so the desktop transcript and the terminal
 * agree on which calls are reasoning and on what text they show.
 */

/** Every accepted spelling of the tool, lowercased. */
export const THINK_TOOL_NAMES = new Set(["deep_think", "think", "think_tool", "deep_thinking"]);

/** Whether a tool name refers to the Thinking Tool scratchpad. */
export function isThinkTool(name: unknown): boolean {
  return THINK_TOOL_NAMES.has(String(name ?? "").trim().toLowerCase());
}

function clean(value: unknown): string {
  if (value === null || value === undefined) return "";
  if (Array.isArray(value)) {
    return value
      .map((item) => String(item ?? "").trim())
      .filter(Boolean)
      .join("\n")
      .trim();
  }
  return String(value).trim();
}

/**
 * Renderable thinking text from a `deep_think` call's arguments.
 *
 * Accepts the raw JSON string the model streamed as well as a parsed object, so
 * a half-finished call still shows something instead of collapsing to nothing.
 */
export function thinkToolText(args: unknown): string {
  let payload: unknown = args;
  if (typeof payload === "string") {
    const trimmed = payload.trim();
    if (!trimmed) return "";
    try {
      payload = JSON.parse(trimmed);
    } catch {
      // Arguments still streaming in, or not JSON at all -- show them verbatim.
      return trimmed;
    }
  }
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) return "";
  const record = payload as Record<string, unknown>;
  const topic = clean(record.topic);
  const thought = clean(record.thought);
  const nextStep = clean(record.next_step);
  const sections: string[] = [];
  if (topic) sections.push(`**${topic}**`);
  if (thought) sections.push(thought);
  if (nextStep) sections.push(`Next: ${nextStep}`);
  return sections.join("\n\n").trim();
}

File: ReverieCli-py/reverie/thinking_tool.py
"""Shared vocabulary for the experimental Thinking Tool.

The tool implementation, the terminal renderer, and the GUI bridge all need the
same tool names and the same "what text do we actually show" rule.  Keeping them
in a dependency-free leaf module lets the CLI display layer use them without
importing the whole ``reverie.tools`` package at startup.
"""

from __future__ import annotations

from typing import Any, Dict, Optional


#: Primary tool name advertised to the model.
THINK_TOOL_NAME = "deep_think"

#: Aliases the model may use for the same call.
THINK_TOOL_ALIASES: tuple[str, ...] = ("think", "think_tool", "deep_thinking")

#: Tag the terminal renders this content under.
THINK_TOOL_TAG = "think_tool"

#: Every accepted spelling, lowercased, for cheap membership tests.
THINK_TOOL_NAMES = frozenset(
    {THINK_TOOL_NAME.lower()} | {alias.lower() for alias in THINK_TOOL_ALIASES}
)


def is_think_tool(tool_name: Any) -> bool:
    """Whether a tool name refers to the Thinking Tool scratchpad."""
    return str(tool_name or "").strip().lower() in THINK_TOOL_NAMES


def clean_think_argument(value: Any) -> str:
    """Normalize an argument a model may send as any JSON scalar or list."""
    if value is None:
        return ""
    if isinstance(value, (list, tuple)):
        value = "\n".join(str(item) for item in value if str(item or "").strip())
    elif not isinstance(value, str):
        value = str(value)
    return value.strip()


def extract_think_tool_text(arguments: Optional[Dict[str, Any]]) -> str:
    """Return the renderable thinking text from a ``deep_think`` call's arguments.

    Shared by the terminal renderer and the GUI bridge so both show the same
    body, including for partially-streamed or malformed argument payloads.
    """
    if not isinstance(arguments, dict):
        return ""

    thought = clean_think_argument(arguments.get("thought"))
    topic = clean_think_argument(arguments.get("topic"))
    next_step = clean_think_argument(arguments.get("next_step"))

    sections: list[str] = []
    if topic:
        sections.append(f"**{topic}**")
    if thought:
        sections.append(thought)
    if next_step:
        sections.append(f"Next: {next_step}")
    return "\n\n".join(sections).strip()


__all__ = [
    "THINK_TOOL_ALIASES",
    "THINK_TOOL_NAME",
    "THINK_TOOL_NAMES",
    "THINK_TOOL_TAG",
    "clean_think_argument",
    "extract_think_tool_text",
    "is_think_tool",
]
